vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+        &mut self,
+        writer: &mut queue::WriteBatch<'_, M>,
+        packet_type: OutgoingPacketType<'_>,
+        transaction_id: u64,
+        operation: storvsp_protocol::Operation,
+        status: storvsp_protocol::NtStatus,
+        payload: &[u8],
+    ) -> Result<(), StorvscError> {
+        let header = storvsp_protocol::Packet {
+            operation,
+            flags: 0,
+            status,
+        };
+
+        // storvsp limits the size of the completion packet to the size of the request packet,
+        // so we need to pad the payload to the maximum size to ensure we get a complete response.
+        // The maximum size includes header + payload.
+        let len = size_of_val(&header) + size_of_val(payload);
+        let padding = [0; storvsp_protocol::SCSI_REQUEST_LEN_MAX];
+        let (payload_bytes, padding_bytes) = if len > storvsp_protocol::SCSI_REQUEST_LEN_MAX {
+            (
+                &payload[..storvsp_protocol::SCSI_REQUEST_LEN_MAX - size_of_val(&header)],
+                &[][..],
+            )
+        } else {
+            (
+                payload,
+                &padding[..storvsp_protocol::SCSI_REQUEST_LEN_MAX - len],
+            )
+        };
+        assert_eq!(
+            size_of_val(&header) + payload_bytes.len() + padding_bytes.len(),
+            storvsp_protocol::SCSI_REQUEST_LEN_MAX
+        );
+        writer
+            .try_write(&OutgoingPacket {
+                transaction_id,
+                packet_type,
+                payload: &[header.as_bytes(), payload_bytes, padding_bytes],
+            })
+            .map_err(|err| match err {
+                queue::TryWriteError::Full(_) => StorvscError::NotEnoughSpace,
+                queue::TryWriteError::Queue(err) => StorvscError::Queue(err),
+            })
+    }
+
+    async fn send_packet_and_expect_completion<
+        M: RingMem,
+        P: IntoBytes + Immutable + KnownLayout,
+    >(
+        &mut self,
+        queue: &mut Queue<M>,
+        operation: storvsp_protocol::Operation,
+        payload: &P,
+    ) -> Result<StorvscCompletionPacket, StorvscError> {
+        let (mut reader, mut writer) = queue.split();
+        let transaction_id = self.get_next_transaction_id();
+        self.send_packet(
+            &mut writer,
+            operation,
+            storvsp_protocol::NtStatus::SUCCESS,
+            transaction_id,
+            payload,
+        )?;
+        // Wait for completion
+        let completion = match self.next_packet(&mut reader).await? {
+            Packet::Completion(packet) => Ok(packet),
+            //Packet::Data(_) => Err(StorvscError::PacketError(PacketError::InvalidPacketType)),
+        }?;
+        expect_success(
+            expect_transaction_id(completion, transaction_id).map_err(StorvscError::PacketError)?,
+        )
+        .map_err(StorvscError::PacketError)
+    }
+}
+
+enum Packet {
+    Completion(StorvscCompletionPacket),
+    //Data(StorvscDataPacket),
```

Comment:
I believe this case will be necessary for supporting hot add/remove.

Reply:
What would the protocol operations be for hot add/remove?

Reply:
The driver get an async notification message that something changed, and then it's supposed to rescan the SCSI bus (if even needs to--in the OpenHCL case, we know what LUN to expect, so we can pretty much just silently drop the notification).

Reply:
After doing nothing, then we need to send a completion for the data packet, correct?

Reply:
Which commands do we need to support receiving in storvsc? I imagine ENUMERATE_BUS is among them, but are there any others?

Reply:
Correct, `ENUMERATE_BUS` is the key command. One thing I noticed is that this is a good implementation of the storvsc protocol, but it doesn't seem that we have the machinery in place to hook this up. I'll start a separate comment for that.

Reply:
I believe this is resolved now since we support `ENUMERATE_BUS` as a no-op